### PR TITLE
Refactor for scraped

### PR DESCRIPTION
**Pull request by ondenman**
This is the first step in moving the scraper to scraped.

The output of this PR matches the output in master.

Further changes to be made after this PR:

* Fix the image field so the correct image is captured for all members (everypolitician/everypolitician-data#33716)
* Split multiple phone numbers and emails (everypolitician/everypolitician-data#33726)

### CODE REVIEW

File: scraper.rb [hunk 1]
```diff
@@ -5,107 +5,25 @@
 require 'pry'
 require 'scraped'
 require 'scraperwiki'
+require 'nokogiri'
+require 'scraped_page_archive/open-uri'
+require 'date'
+require 'scraped'
```
Comment (tmtmtmtm) [inline scraper.rb]: I'm assuming you've reintroduced these by mistake?

File: scraper.rb [hunk 2]
```diff
@@ -5,107 +5,25 @@
...
-    gender:           gender,
-    source:           url.to_s,
-  }
-
-  data = data.merge(area_info)
-
+list_url = 'http://www.legco.gov.hk/general/english/members/yr16-20/biographies.htm'
+(scrape list_url => MembersPage).member_urls.each do |url|
+  data = (scrape url => MemberPage).to_h.merge(term: 6)
   ScraperWiki.save_sqlite([:id], data)
```
Comment (tmtmtmtm) [inline scraper.rb]: our normal practice here is to do this `save` in one shot after building up all the data, rather than within the loop.

File: scraper.rb [hunk 3]
```diff
@@ -5,107 +5,25 @@
...
+  # puts data.reject { |k, v| v.to_s.empty? }.sort_by { |k, v| k }.to_h
 end
-
-ScraperWiki.sqliteexecute('DROP TABLE data') rescue nil
```
Comment (tmtmtmtm) [inline scraper.rb]: why have you removed the `DROP TABLE`?

File: test/data/suk-yee.yml [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+:to_h:
+  :id: yr16-20/ilsy
+  :name: Regina IP LAU Suk-yee
+  :honorific_prefix: Hon Mrs
+  :honorific_suffix: GBS, JP
+  :gender: female
+  :faction: New People's Party
+  :email: iplau@reginaip.hk
+  :website: http://www.reginaip.hk
+  :phone: 2537 3267 / 2537 3265
```
Comment (tmtmtmtm) [inline test/data/suk-yee.yml]: I think you should add an explicit `TODO` note here that this isn't what we want

File: lib/member_page.rb [hunk 5]
```diff
@@ -0,0 +1,108 @@
+# frozen_string_literal: true
+
+require 'scraped'
+require 'pry'
```
Comment (tmtmtmtm) [inline lib/member_page.rb]: I don't think you want/need this to be here.

File: lib/member_page.rb [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+# This class represents the profile page of a given member
+class MemberPage < Scraped::HTML
+  decorator Scraped::Response::Decorator::CleanUrls
+
+  field :id do
+    url.gsub('.htm', '').split('members/').last
```
Comment (tmtmtmtm) [inline lib/member_page.rb]: This doesn't really seem like how you want to doing this.

`gsub` implies that you're expecting that there might be more than one instance of the '.htm', and generally isn't really suitable for this sort of thing anyway. Normally you'd just want `File.basename(url, '.htm')` here, though this case is a little more awkward because of wanting the 'yrDD' part too.

Building both up here in one place would be a little ugly, but the first part of that is really the `term` field anyway (even if `scraper.rb` is going to want that to be '6' instead, the class should really be exposing what the site claims), so if that already exists, then this becomes something more like:

```ruby
  field :id do
    '%s/%s' % [term, File.basename(url, '.htm')]
  end

  field :term do
    File.dirname(url).split('/').last
  end
```

File: lib/member_page.rb [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+  end
+
+  field :name do
+    name_parts.first.to_s.gsub(Regexp.union(titles << '.'), '').tidy
```
Comment (tmtmtmtm) [inline lib/member_page.rb]: it's not really clear what this `titles << '.'` is doing, but the tests pass without it, so if it's important I think you need an extra test case.

This approach also seems flawed, in that it's going to fail if anyone ever has a name that contains one of the prefixes (e.g. "Draper").

More generally, this is already a solved problem in other scrapers — why are you reinventing a new solution to it rather than reusing an existing approach?

File: lib/member_page.rb [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+  end
+
+  field :honorific_prefix do
+    titles.select { |prefix| name_parts.first.to_s.include? prefix }.join(' ')
+  end
+
+  field :honorific_suffix do
+    name_parts[1..-1].map(&:tidy).join(', ')
```
Comment (tmtmtmtm) [inline lib/member_page.rb]: I suspect you want the `tidy` to already have been applied to the `name_parts` before now.

`name_parts.drop(1)` would also be a lot clearer than `name_parts[1..-1]`

File: lib/member_page.rb [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+  end
+
+  field :gender do
+    return 'female' if honorific_prefix.include?('Mrs')
+  end
+
+  field :faction do
+    f = bio.xpath('//p[contains(.,"Political affiliation")]/'\
+                  'following-sibling::ul[not(position() > 1)]/li/text()')
```
Comment (tmtmtmtm) [inline lib/member_page.rb]: That's worth factoring out to a separate method, rather than doing both the finding and the processing here.

File: lib/member_page.rb [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+  def bio
+    noko.css('div#container div')
+  end
+
+  def non_party_groups
+    [
+      'Kowloon West New Dynamic',
+      'New Territories Association of Societies',
+      'April Fifth Action',
+    ]
```
Comment (tmtmtmtm) [inline lib/member_page.rb]: it's a fairly small list, so it's not a massive issue here, but it's worth getting into the habit of using a Set instead of an Array when a list exists solely to be scanned, so that lookup is O(1) instead of O(n)

File: lib/member_page.rb [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+
+  field :faction do
+    f = bio.xpath('//p[contains(.,"Political affiliation")]/'\
+                  'following-sibling::ul[not(position() > 1)]/li/text()')
+    return 'Independent' if f.empty?
+
+    # Some member pages list more than one group affiliation for that member
+    # Here, we remove affiliations with known non-party groups
+    f.map(&:to_s).map(&:tidy).find do |party|
+      !non_party_groups.to_s.include? party
```
Comment (tmtmtmtm) [inline lib/member_page.rb]: a `find` with a `not` like this is a little bit awkward, and also disguises slightly the case of what's happening if there's more than one non-party-group party. 

I think this might be a little bit clearer and more explicit as `.reject { … }.first`

File: lib/member_page.rb [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+
+  field :img do
+    # TODO: incorrect image being captured for 'WONG Ting-kwong'
+    # Change line to: bio.at_css('img/@src').to_s
+    bio.css('img/@src').last.to_s
+  end
+
+  field :area do
+    # splitting here by en-dash (not hyphen)
+    area_parts.last.split('–').last.tidy
```
Comment (tmtmtmtm) [inline lib/member_page.rb]: Changing this to a hyphen doesn't cause a test failure, so either it's unnecessary or you want an extra test case…

BTW: rather than having to draw attention to the specific character with a comment, perhaps it might be clearer to use a unicode string explicitly: `.split("\u{2013}")`?

File: lib/members_page.rb [hunk 13]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+require 'scraped'
+
+# This class represents a page listing members of the given legislature
+class MembersPage < Scraped::HTML
+  decorator Scraped::Response::Decorator::CleanUrls
+
+  field :member_urls do
+    noko.css('.bio-member-detail-1 a/@href').map(&:to_s)
```
Comment (tmtmtmtm) [inline lib/members_page.rb]: We usually use `.text` rather than `.to_s` there, so unless there's a reason why that doesn't work here, it's better to be consistent.

**Review by tmtmtmtm — CHANGES_REQUESTED** [review]
This ended up being much more of a rewrite than I was expecting. In general you've done a pretty good job of that, but there are a few too many places where you're doing things in a non-standard way and it's difficult to know if that's deliberate/needed.

File: lib/member_page.rb (hunk 6 above)
Comment (tmtmtmtm) [reply lib/member_page.rb]: Looking at this again now, perhaps a join would be clearer on the `:id`:

```ruby
  field :id do
    [term, File.basename(url, '.htm')].join('/')
  end
```

File: lib/member_page.rb [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+  field :honorific_suffix do
+    name_parts.drop(1).map(&:tidy).join(', ')
+  end
+
+  field :gender do
+    return 'female' if honorific_prefix.include?('Mrs')
+  end
+
+  field :faction do
+    return 'Independent' if (affiliation = political_affiliation).empty?
```
Comment (tmtmtmtm) [inline lib/member_page.rb]: Creating an extra `affiliation` variable here is a little clumsy, and doesn't really buy us anything, other than the cost of an extra method call to `political_affiliation` (if that method in turn were slow then we could memoise it, but as it's just an XPath lookup, then I don't think there's any need for that either). I'd just leave this as `if political_affiliation.empty?`

File: lib/member_page.rb [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+    # Some member pages list more than one group affiliation for that member
+    # Here, we remove affiliations with known non-party groups
+    affiliation.map(&:to_s).map(&:tidy).reject do |party|
+      non_party_groups.to_s.include? party
```
Comment (tmtmtmtm) [inline lib/member_page.rb]: Why are you casting `non_party_groups` to a string here? The idea of it being a `Set` is that you have an O(1) lookup directly into it…

File: lib/member_page.rb [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+  field :id do
+    [term, File.basename(url, '.htm')].join('/')
+  end
+
+  field :term do
+    File.dirname(url).split('/').last
+  end
+
+  field :name do
+    name_parts.first.to_s.gsub(Regexp.union(titles << '.'), '').tidy
```
Comment (tmtmtmtm) [inline lib/member_page.rb]: My previous concerns here still apply…

**Review by tmtmtmtm — CHANGES_REQUESTED** [review]

File: lib/member_page.rb [hunk 17]
```diff
@@ -31,11 +31,11 @@ class MemberPage < Scraped::HTML
...
   field :faction do
-    return 'Independent' if (affiliation = political_affiliation).empty?
+    return 'Independent' if political_affiliation.empty?
     # Some member pages list more than one group affiliation for that member
     # Here, we remove affiliations with known non-party groups
-    affiliation.map(&:to_s).map(&:tidy).reject do |party|
-      non_party_groups.to_s.include? party
+    political_affiliation.map(&:to_s).map(&:tidy).reject do |party|
+      non_party_groups.include? party
     end.first
```
Comment (tmtmtmtm) [inline lib/member_page.rb]: Ditto the above comment for `list1 - list2` vs `list1.reject { |e| list2.include? e }`

NB: this removal/subtraction appears to be untested. If I drop the `reject` part of this, and make this simply ```political_affiliation.map(&:to_s).map(&:tidy).first```, the tests still pass.

File: lib/member_page.rb [hunk 18]
```diff
@@ -15,7 +15,7 @@ class MemberPage < Scraped::HTML
   end
 
   field :name do
-    name_parts.first.to_s.gsub(Regexp.union(titles << '.'), '').tidy
+    name_parts.first.split.reject { |a| titles.include? a }.map(&:tidy).join(' ')
```
Comment (tmtmtmtm) [inline lib/member_page.rb]: the logic here is quite odd.

1. Why are you calling `tidy` _after_ the `reject`? If the parts aren't already suitable tidied, then they won't be correctly filtered out. If they _are_, then why re-tidy them again? What's that protecting against?

2. `name_parts.first.split` is a little hard to follow, possibly as `name_parts` doesn't really convey what the data is you're dealing with (and worse makes it sound like it's already the parts of the name, in which case why are we splitting them up again). I suspect it would be better to factor this out as another method, and make sure both have suitably descriptive names.

3. Iterating over a list to remove, one by one, the members of another list, is quite long-winded, confusing, and inefficient. `list1 - list2` is much simpler and clearer.

**Review by tmtmtmtm — CHANGES_REQUESTED** [review]

**Comment by tmtmtmtm** [comment]
@ondenman this doesn't autosquash cleanly. Can you squash it down to a series of clean commits?

**Comment by ondenman** [comment]
I've squashed the commits down and have opened a separate PR (https://github.com/everypolitician-scrapers/hong_kong_legislative_council_members/pull/8) to setup the test framework. I will move the commits to a new branch so Travis runs the regression tests.

**Comment by ondenman** [comment]
Closing in favour of: https://github.com/everypolitician-scrapers/hong_kong_legislative_council_members/compare/use-scraped?expand=1
